### PR TITLE
HPCC-29848 Optimize the CJoinGroup allocations

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman this did seem to make a noticable difference on stresstext.ecl, but that isn't very typical so I wouldn't expect as large an affect in general

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -26450,6 +26453,7 @@ class CRoxieServerKeyedJoinBase : public CRoxieServerActivity, implements IRecor
     virtual void reset()
     {
         CRoxieServerActivity::reset();
+        joinGroupAllocator->emptyCache();
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp:26456]: Not sure what the benefit of this is. If it's in a child query, the cache would be useful. If it's not, emptying the cache here is pointless? At least based on my memory/understanding of how this works...

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -26354,6 +26355,8 @@ class CRoxieServerKeyedJoinBase : public CRoxieServerActivity, implements IRecor
         indexReadInput = NULL;
         rootIndex = NULL;
         atmostsTriggered = 0;
+        unsigned allocatorFlags = roxiemem::RHFblocked;//roxiemem::RHFpacked; // MORE? unique? blocked?
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp:26358]: Should this be controlled via a hint? Or should you just make your mind up? Comment could use some cleanup at least...

**Review by richardkchapman — COMMENTED** [review]
Couple of comments

File: roxie/ccd/ccdserver.cpp (hunk 1 above)
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp:26456]: This releases any rows that have been allocated and haven't been used.  Without it you will get complains about leaked rows.

File: roxie/ccd/ccdserver.cpp (hunk 2 above)
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp:26358]: I agree.  I think blocked is sufficient - it has fewer side effects than packed, which would create a unique heap.  Assuming I am correct that alloc() is only called from a single thread...

**Comment by ghalliday** [comment]
Cleaned up comment.  It would make sense to review all allocators, but something to do in a separate request.

**Review by richardkchapman — APPROVED** [review]
Approved, but should probably understand the smoke test failure

**Pull request merged**